First pass at creating an Elasticsearch index. DSDEGP-1232

kshakir opened the pull request:
### Description

Added creation of a readgroups bam index.
While adding a spec for creation of the ES index, added specs for the services.
The HTTP server now waits to startup until the ES indexes are populated.
The server startup and shutdown are recorded internally with additional statuses beyond "OK".
Bumped up dependency versions while addressing an Akka-HTTP warning with 10.0.0.5.
While updating dependency versions, broke dependencies and other sections of build.sbt into separate project files.
DSDEGP-1232

----

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below.

- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [x] **Submitter**: Ensure that you have added or updated tests
- [x] **Submitter**: If you're adding/switching libraries or otherwise changing the design, update the [design documentation](https://broadinstitute.atlassian.net/wiki/pages/viewpage.action?pageId=114531509).
- [x] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * Add notes on what you've tested
* Review cycle:
  * Team may comment on PR at will
  * **Reviewer assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to reviewer** for further feedback
- [x] **Submitter**: Verify all tests go green, including CI tests
- [x] **Submitter**: Squash commits and merge to develop
- [x] **Submitter**: Delete branch after merge
- [x] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [x] **Submitter**: Move the JIRA issue to QA once this checklist is completed

kshakir committed:
Build updates.
Refactored build.sbt into separate files.
Added a clio version docker image label.
Writing clio jar to /app/clio.jar instead of using a version in the file name.
Added a few more scalac checks found online.
Bumped dependencies, especially Akka-HTTP to 10.0.6.
Explicitly load the ClioIntegrationTestPlugin.

kshakir committed:
Refactored server startup.
A new ServerStatusDAO tracks the server status from NotStarted to Shutdown.
Clio waits for a configurable list of elasticsearch statuses before listening on the http server.
If the server fails to contact elasticsearch, retries a number of times.
When retries fail, the program exits.

kshakir committed:
Added service specs.

pshapiro4broad commented on src/main/scala/org/broadinstitute/clio/dataaccess/CachedServerStatusDAO.scala:
```diff
@@ -0,0 +1,47 @@
+/**
+  * Stores the server status in a synchronized mutable variable.
+  */
+class CachedServerStatusDAO private() extends ServerStatusDAO {
```
@rtitle is currently adding (or done adding?) server status to rawls. It might be worth looking at what he's done, to avoid work duplication, even if we don't have a way to share code across projects.

kshakir replied:
I'll ping Rob. Thanks!

rtitle replied:
Hi - I haven't fully looked at this PR (or even know what this repo is for :)), but my [Rawls PR](https://github.com/broadinstitute/rawls/pull/703) is more about asynchronous monitoring of subsystems to provide health checks. It seems like this is more about lifecycle management? So maybe slightly different use cases, though I'd love to discuss the possibility of sharing code.

As an aside, I've used [Guava service](https://github.com/google/guava/wiki/ServiceExplained) before as a way to manage services which have a lifecycle/state and found it pretty nice to use (granted this was in a Java project).

pshapiro4broad commented on src/main/scala/org/broadinstitute/clio/dataaccess/CachedServerStatusDAO.scala:
```diff
@@ -0,0 +1,47 @@
+    Try {
+      mutex synchronized {
+        import CachedServerStatusDAO.level
+        if (level(currentStatus) < level(status))
```
It seems like you'd want an abstraction here, e.g. `canTransitionTo()` on `ServerStatusInfo`, in case the rule for transitioning status change or become more complicated.

pshapiro4broad commented on src/main/scala/org/broadinstitute/clio/dataaccess/CachedServerStatusDAO.scala:
```diff
@@ -0,0 +1,47 @@
+
+  private def level(status: ServerStatusInfo): Int = {
+    status match {
+      case ServerStatusInfo.NotStarted => 0
```
Would it be possible to store this information directly in `ServerStatusInfo`, e.g. as per-instance state? Doing that avoids the need to handle the "unknown" case and prevents the situation where a user adds a new status and forgets to add it here.

kshakir replied:
I wrestled with this one a bit. Ultimately I decided to follow the firecloud-orchestration [DEVNOTES](https://github.com/broadinstitute/firecloud-orchestration/blob/63dcb0be65b2fe69234b0e139c4ce6e57bf90087/DEVNOTES.md), that are copied into the clio [wiki](https://broadinstitute.atlassian.net/wiki/pages/viewpage.action?pageId=114531509#Clio(MetadataManager)-PackageStructureandSingleResponsibilities).

> - model
>     -  model classes. Case classes if possible. Little-to-no logic implementations; things like toString allowed

Thus the `ServerStatusInfo` model objects passed around do not contain any business logic, not even the ability to transition from one "level" to another.

pshapiro4broad reviewed (commented):
Would it be worth creating an abstraction layer that separates elasticsearch from our code? That way you can separate the functionality that elasticsearch provides (which is the abstraction layer), and elasticsearch itself (which is the implementation layer). Part of the reason for doing this is so we can provide a name for the class that describes its function in our code, instead of using the name of a product, which by itself may not have that meaning.

kshakir commented:
Re:

> Would it be worth creating an abstraction layer that separates elasticsearch from our code?

As of the latest commit, all ES specific implementation details are _only_ in the HttpElasticsearchDAO. The abstract interface ElasticsearchDAO uses concepts from ES, but these could easily be implemented for any other another data storage type, in theory. In memory, RDBMS, filesystems, etc. would all work.

However, there are still concepts in the model of an "Elasticsearch specific index". For example, in addition to the index-name, there is also an [index-type](https://www.elastic.co/blog/index-vs-type).

That said, all this could be more clearly defined in Clio. I am notoriously bad an naming, and happy to discuss alternate naming strategies. FYI- I chose the names peeking at the Firecloud Orchestration code. They seem to name the [abstract DAOs](https://github.com/broadinstitute/firecloud-orchestration/tree/b42c1bf37f162c469fc38cc5d69cbbd401610663/src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess) based on the type of client, not the function of the DAO.

kshakir committed:
First pass at creating an Elasticsearch index.
On server startup, once connected to elasticsearch, the indices are created or updated.
A ReflectionIndexDefiner reads the fields from a case class to create the fields.
Added an example ElasticsearchIdExtractor, currently unwired.

kshakir commented:
Verbal approval received. Leaving commits separated by what they contribute to the codebase.
